src/d3a/models/state.py

```diff
@@ -43,131 +46,249 @@
 # - If a device has no state, maybe it doesn't need its own appliance class either
 
 
-class PVState:
+class StateInterface(metaclass=ABCMeta):
+    """Interface containing methods that need to be defined by each State class."""
+
+    @abstractmethod
+    def get_state(self) -> Dict:
+        """Return the current state of the device."""
+        return {}
+
+    @abstractmethod
+    def restore_state(self, state_dict: Dict):
+        pass
+
+    @abstractmethod
+    def delete_past_state_values(self, time_slot):
+        pass
+
+    def __str__(self):
+        return self.__class__.__name__
+
+
+class ConsumptionState(StateInterface):
+    """State for devices that can consume energy."""
+
     def __init__(self):
+        super().__init__()
+        # Energy that the load wants to consume (given by the profile or live energy requirements)
+        self._desired_energy_Wh = {}
+        # Energy that the load needs to consume. It's reduced when new energy is bought
+        self._energy_requirement_Wh = {}
+        self._total_energy_demanded_Wh = 0
+
+    def get_state(self) -> Dict:
+        """Return the current state of the device. Extends super implementation."""
+        state = super().get_state()
```

Comment:
If the `get_state` is to stay abstract this line can be safely removed

Reply:
This line is used to allow the collaboration between mixins (when using multiple inheritance). That's also the reason why the abstract `get_state()` method in `StateInterface` is returning `{}`. Removing that would break the code that follows (that expects to handle a dict object).

The method resolution order for the `HomeMeterState` is the following:
```python
HomeMeterState.__mro__
(<class 'HomeMeterState'>, <class 'ConsumptionState'>, <class 'ProductionState'>, <class 'StateInterface'>, <class 'object'>)
```
So calling `super().get_state()` would call the `get_state()` method on all siblings/parents in the following order:
```
HomeMeterState -> ConsumptionState -> ProductionState -> StateInterface
```
If the `StateInterface.get_state()` returns `None`, the `ProductionState.get_state()` method will fail. We also want to allow using mixins flexibly, so this shouldn't be fixed in every individual state that implements `StateInterface`, but rather at the source.
I see some alternatives:
1. Do not make the method abstract (but that would sort of defeat the purpose of the interface class)
2. Check for `None` in every class that implements the interface (`ConsumptionState` and `ProductionState`).
3. Leave it as it is (use `@abstractmethod` but also provide implementation)

What do you think? Do you have any other ideas?

Reply:
I see, the solution #2 looks the best imo

Reply:
I would actually personally prefer to leave it as it is. Reason for that is that I predict that these classes will be composed in ways that we cannot envision now (for instance, I can imagine once we add power flow to have a power state, which would then convert the MRO to HomeMeterState -> ConsumptionState -> ProductionState -> PowerState -> StateInterface) and this would make it inconvenient to have to check for None for every implemented state class. Furthermore, default implementations in abstractmethods / interfaces are even allowed in Java nowadays, therefore I would not consider this an antipattern.